### PR TITLE
Set up for translations

### LINKED ISSUE #566: Translations: mark up strings to be translated
Decide which strings to translate, if we can only do a subset of them. Mark them up with the gettext `_(...)` syntax using a script.

### LINKED ISSUE #568: Translations: Add translation infrastructure
This will involve:
    - making a way for UI "choose a language" components to communicate the chosen language to invest and reload
        - new endpoint for the workbench
        - ??? for Qt if we do that
    - calling `gettext.translation(...).install()` somewhere (cli.py?) 
    - creating directories in the expected structure to store our translation .po files
    - calling `pybabel compile` as part of the install/build process (converts human-readable .po files to binary .mo files, which are used to look up strings at runtime)

### PR DESCRIPTION
**Pull request by emlys**
# Description
Fixes #566, #568 

For context see 
* [`src/natcap/invest/internationalization/README.md`](https://github.com/emlys/invest/blob/task/568/src/natcap/invest/internationalization/README.md) 
* Minimal example of `gettext`: https://github.com/emlys/python-gettext-demo
* Design doc: https://docs.google.com/document/d/1hB3ZOYwN-qXYt4mvq7cM5ZTikPqnD00m7W1izom5leM
* PEP: https://docs.google.com/document/d/1m5TpHoVu6uGVQ7Us8BGTf-blmdeaiPIpBZYy1K9duVg
* Flowchart: https://lucid.app/lucidchart/18ad9fcf-8117-463f-b9ed-2cf2a46ffa9d/edit?viewport_loc=-445%2C378%2C2605%2C1428%2C0_0&invitationId=inv_91a6df63-6199-4453-a298-ca42f08d2880 (experimenting with flowcharts to accompany plain text documentation. this was helpful for me thinking about the problem. let me know if it is/isn't helpful to you too!)

This was already quite large so I split out user-provided translations into a separate issue #723 

## Changes
* Add a `language` argument to the CLI. This affects `invest validate` (validation messages), `invest getspec` (arg names and descriptions), and `invest list` (model names).
* Accept a `language` query parameter at UI server endpoints. Paralleling the CLI changes, this affects `validate`, `getspec`, and `models`.
* Wrap all strings to translate in the `_(...)` function

I also consolidated the validation messages in the top level of `validation.py` so that they aren't duplicated throughout the tests. Seemed like a good time to do that, so that tests are independent of translation (though our tests will still run in English).

# Checklist
- [x] Updated HISTORY.rst (if these changes are user-facing)
~- [ ] Updated the user's guide (if needed)~

### CODE REVIEW

**Comment by emlys** [comment]
@phargogh @davemfish @dcdenu4 , I'm calling it ready. The mac binary build failure looks like some intermittent github actions issue to me.

**Review by phargogh — CHANGES_REQUESTED** [review]
Man, this is quite a reshuffle!  Thanks for all of your work on this, and also clearly thinking through to how we'll need to maintain it long-term (like by adding the `README.md` in the i18n folder).  I had a few comments, questions and suggestions, and please let me know what you think and/or if you'd like to talk through any of them!

File: HISTORY.rst [hunk 1]
```diff
@@ -35,9 +35,14 @@
...
-  Unreleased Changes
-  ------------------
+
+Unreleased Changes
+------------------
+* General:
+    * Add a ``--language`` argument to the command-line interface, which will
+      translate model names, specs, and validation messages.
+    * Accept a ``language`` query parameter at the UI server endpoints, which
+      will translate model names, specs, and validation messages.
```
[the remark this answers is not in the document]
Comment (emlys) [reply HISTORY.rst:45]: I changed it to say `Unreleased Changes (3.10)` - is that what you mean?

File: tests/test_habitat_quality.py [hunk 2]
```diff
@@ -1824,9 +1825,8 @@ def test_habitat_quality_duplicate_threat_path(self):
         actual_message = str(cm.exception)
         # assert that a duplicate error message was raised
         self.assertTrue(
-            'Threat paths cannot be the same and must have '
-            'unique absolute filepaths.' in
-            actual_message, actual_message)
+            habitat_quality.DUPLICATE_PATHS_MSG in actual_message,
+            actual_message)
```
[the remark this answers is not in the document]
Comment (emlys) [reply tests/test_habitat_quality.py:1829]: Oh neat! Thanks for pointing that out.

[the remark this answers is not in the document]
Comment (emlys) [reply tests/test_habitat_quality.py]: I replaced all `assertTrue(x in y)` in the HQ tests with `assertIn(x, y)`.

File: requirements.txt [hunk 3]
```diff
@@ -25,3 +25,4 @@ psutil>=5.6.6
 chardet>=3.0.4
 openpyxl
 pint
+Babel
```
[the remark this answers is not in the document]
Comment (emlys) [reply requirements.txt]: good catch!

File: src/natcap/invest/cli.py [hunk 4]
```diff
@@ -377,6 +380,8 @@ def main(user_args=None):
 
         model_module = importlib.import_module(
             name=parsed_datastack.model_name)
+        importlib.reload(importlib.import_module(name='natcap.invest.validation'))
+        importlib.reload(model_module)
```
[the remark this answers is not in the document]
Comment (emlys) [reply src/natcap/invest/cli.py:384]: Now that you point that out, I don't think we need either of these reloads. I added them to be safe, but in the CLI, `main()` will just run one time through and exit. We can be sure that neither `validation` nor the model module have been imported before. So it's safe to simply import them without reloading -  they will execute on the first import.

File: src/natcap/invest/cli.py [hunk 5]
```diff
@@ -253,6 +246,12 @@ def main(user_args=None):
         action='store_const', const=logging.DEBUG,
         help='Enable debug logging. Alias for -vvv')
 
+    parser.add_argument(
+        '-L', '--language', default='en', choices=['en', 'es'],
+        help=('Choose a language. Model specs, names, and validation messages '
+              'will be translated. Log messages are not translated. Value '
+              'should be an ISO 639-1 language code.'))
```
[the remark this answers is not in the document]
Comment (emlys) [reply src/natcap/invest/cli.py]: I found a neat feature of `babel` for this! They provide standard locale data from the [Common Locale Data Repository](https://cldr.unicode.org/), including the name of each language in that language. Now the contents of `internationalization/locales` will be the single source of truth for which languages are supported, and we can easily get their names too.

File: requirements.txt (hunk 3 above)
[the remark this answers is not in the document]
Comment (emlys) [reply requirements.txt:28]: I reverted this back to `requirements.txt` since I'm now using `babel` to get the localized language names.

File: src/natcap/invest/__init__.py [hunk 6]
```diff
@@ -30,176 +46,188 @@ class _MODELMETA:
...
+        model_title=_('Urban Cooling'),
         pyname='natcap.invest.urban_cooling_model',
         gui='urban_cooling_model.UrbanCoolingModel',
         userguide='urban_cooling_model.html',
         aliases=('ucm',)),
 }
 
 
+def install_language(language_code):
+    """Globally install the _() function for the requested language."""
```
[the remark this answers is not in the document]
Comment (emlys) [reply src/natcap/invest/__init__.py:229]: Good idea, I added a `ValueError` if the language code isn't supported. You're right, the fallback applies to each string individually. That's helpful so that we can use an incomplete translation set if necessary.

File: setup.py [hunk 7]
```diff
@@ -37,6 +39,18 @@
...
+# the MO files are included as package data
+locale_dir = os.path.abspath(os.path.join(
+    os.path.dirname(__file__), 'src/natcap/invest/internationalization/locales'))
+for locale in os.listdir(locale_dir):
+    subprocess.run([
+        'pybabel',
+        'compile',
+        '--input-file', f'{locale_dir}/{locale}/LC_MESSAGES/messages.po',
+        '--output-file', f'{locale_dir}/{locale}/LC_MESSAGES/messages.mo'])
+
```
[the remark this answers is not in the document]
Comment (emlys) [reply setup.py]: I was surprised that babel doesn't seem to provide a python API equivalent to the CLI options, hence the `subprocess` call. They do provide setup.py commands - with babel installed, you can do `python setup.py compile_catalog` for instance. But that's becoming obsolete.

You're right doesn't need to run each time `setup.py` is executed. I added a custom `build_py` command that takes care of this. It's like the OPAL example but subclasses the existing `build_py`.

**Comment by emlys** [comment]
@phargogh, I ended up leaving the `reload`s in `cli.py`, even though they're not strictly necessary. It was a lot easier than trying to either perfectly reset the global context between tests, or run each test in its own process. And it does prevent the possibility of problems if we ever import from `cli.py` - with `reload`, we aren't assuming that a module hasn't been imported before.

Something's wrong with the PR diff, it shows 66 files changed but some of those changes were already merged into `release/3.10`, for example the changes shown for `.github/workflows/binary-applications.yml`.

File: HISTORY.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (phargogh) [reply HISTORY.rst]: On the version I was looking at earlier, I think there was a second "Unreleased Changes" or "3.10" section, or something like that ... but that looks like it isn't the case any longer.  So this is resolved!

File: HISTORY.rst [hunk 8]
```diff
@@ -56,7 +60,6 @@ Unreleased Changes (3.10)
       abundance in current/future LULC and therefore less rarity; values
       of 0.5 indicate same abundance between baseline and current/future
       LULC; values 0.5 to 1 indicate less abundance in current/future LULC
-      and therefore higher rarity.
```
Comment (phargogh) [inline HISTORY.rst:59]: Is this removal intentional?

File: src/natcap/invest/__init__.py [hunk 9]
```diff
@@ -30,176 +55,199 @@ class _MODELMETA:
...
+    Args:
+        language_code (str): ISO 639-1 locale code for a language supported
+            by invest
+
+    Returns:
+        None
+    """
+    if language_code not in LOCALES:
+        raise ValueError(
+            f"language '{language_code}' is not supported by InVEST.")
```
Comment (phargogh) [inline src/natcap/invest/__init__.py:240]: What would you think about adding in the supported language codes as well to this error message?

**Review by phargogh — CHANGES_REQUESTED** [review]
Just two minor comments, and otherwise I think this is ready to go!

File: HISTORY.rst (hunk 8 above)
Comment (emlys) [reply HISTORY.rst:67]: No, not sure how that happened!

File: src/natcap/invest/__init__.py (hunk 9 above)
Comment (emlys) [reply src/natcap/invest/__init__.py]: Sure!

**Comment by emlys** [comment]
Please also see https://github.com/emlys/invest/pull/1/ for the `option_string` changes.

**Comment by phargogh** [comment]
@emlys I just took a look at https://github.com/emlys/invest/pull/1 and had a couple minor comments there, so how would you like to proceed?  Shall we wait to merge that PR in to this branch and then we'll continue here?

I also see that there's a failing mac build, but that looks like a Qt error with plugins on mac and something we'll have to resolve, but I don't think it's due to this PR and so we can fix it in a separate issue if it persists.

**Comment by emlys** [comment]
@phargogh thanks, yes let's merge emlys/invest#1 first.

**Comment by phargogh** [comment]
OK @emlys the PR over on your fork is good to go, but I think you'll need to click merge over there since I can't (and shouldn't) be able to push to your fork! 😄

**Comment by emlys** [comment]
The other PR is merged! This should be good to go

**Comment by emlys** [comment]
@phargogh I just resolved the merge conflict on this

**Review by phargogh — APPROVED** [review]
OK!  I think this is good to go!  I see that there's a build failure on mac, but it doesn't immediately appear to be related to the functionality in this PR so I'm OK with merging anyways.
